olp-cpp-sdk-core/src/cache/DefaultCache.cpp

```diff
@@ -195,58 +212,41 @@ boost::any DefaultCache::Get(const std::string& key, const Decoder& decoder) {
     }
   }
 
-  if (mutable_cache_) {
-    auto expiry = GetRemainingExpiryTime(key, *mutable_cache_);
+  bool valid = false;
+  DataPtr result;
+  time_t expiry = 0;
+
+  if (protected_cache_) {
+    std::tie(valid, result, expiry) = GetFromCache(key, *protected_cache_);
+  }
+
+  if (mutable_cache_ && !valid) {
+    std::tie(valid, result, expiry) = GetFromCache(key, *mutable_cache_);
+
     if (expiry <= 0) {
       PurgeDiskItem(key, *mutable_cache_);
-      return boost::any();
     }
+  }
 
-    auto value = mutable_cache_->Get(key);
-    if (value) {
-      auto decoded_item = decoder(value.get());
-      if (memory_cache_) {
-        memory_cache_->Put(key, decoded_item, expiry, value->size());
-      }
-      return decoded_item;
+  if (valid) {
+    auto decoded_item = decoder({result->begin(), result->end()});
```

Comment:
Why do you copy result into the decoder? Seems like a design error to me. If the cost of having a common method is to copy every time one additional time, then i would say: please change back!

Reply:
Redone